File: src/Solutions/Euler070.hs
module Solutions.Euler070 where
import Utils.NumberTheory(totientArrayUpTo)
import Utils.Numeric(digits)
import Utils.Operators((<:>))
import Data.Array.Unboxed((!))

bound = 10000000 :: Int
totient = totientArrayUpTo bound
sameDigits :: Int -> Int -> Bool
sameDigits a b = digits (toInteger a) == digits (toInteger b)
isOk :: Int -> Bool
isOk n = sameDigits n (totient ! n)
ddiv :: Int -> Int -> Double
ddiv a b = (fromIntegral a) / (fromIntegral b)
min' (a, b) (c, d) = if b < d then (a, b) else (c, d)

euler070 :: IO String
euler070 = return $ show $ fst $ foldr (min' . (id <:> (\x -> ddiv x $ totient ! x))) (bound, 2.5) (filter isOk [2..bound])

File: rust/src/solutions/euler070.rs
use crate::utils::number_theory::totient::totient_array;
use crate::utils::numeric::digits::Digits;
use ordered_float::OrderedFloat;
use std::ops::Add;

pub fn euler070() -> String {
    let tot: Vec<usize> = totient_array(10000000);
    (1usize..10000000)
        .filter(|&n| (n + 1).digits_count() == tot[n].digits_count())
        .min_by_key(|&n| OrderedFloat((n as f64 + 1.0) / (tot[n] as f64)))
        .unwrap()
        .add(1)
        .to_string()
}
